Docker compose for testing

pabzm opened the pull request:
I couldn't get the browser tests to run green locally, and felt uneasy about running a downloaded binary with superuser-rights (the IMAP-server) on my machine anyway. So I decided to build a testing setup with docker compose, which depends on external containers for the browser and the IMAP-server and helps to make the environment reproducecable. Additionally it is faster for repeated runs, because it doesn't repeatedly downloads the server-jar, installs the webdriver, etc.

And because more people might find it useful I'm posting this pull request.

We could also rebuild the automated testing workflows to use services (external containers) to speed things up. But before I do that I wanted to know if this gets merged or provokes discussion in the first place.

The container image needs to be built locally, currently. If this gets merged I'd set it up to be built repeatedly on github and to be downloadable.

pabzm committed:
Test setup for local and CI using containers

It uses standalone containers for the greenmail IMAP server and the
standalone browser.
A testrunner image is built in the CI (for `linux/amd64` only, because
Github doesn't support multi-platform building on their default
runners and we don't have our own.)

This setup helps to run the tests (reproduceably) also locally.
Previously, on my machine, they produced varying results.
It also reduces the dependencies for running the browser test.
Local execution only depends on `docker compose`, no other tools
(previously it required `sudo`, `java`, and some more.)

The previous solution should still work, if you want it.

The scripts are stored in a directory called `.ci` to hide them a little
and avoid confusion with the container images from the
`roundcubemail-docker` repo.

pabzm committed:
More ignore-patterns in eslint config

So we don't have to specify them on the command line when we check
codestyle locally.

alecpl commented:
I don't like docker and I have no clue about docker, but...
1. I'm not against using it in CI, as long as possibility to use local/external IMAP server is still there.
2. There's  more or less official roundcube/roundcubemail-docker repository, so we should not add any docker files here, unless they reside in .ci folder.

pabzm commented:
Do I understand correctly that you would merge this if the image is built in https://github.com/roundcube/roundcubemail-docker and we drop the `.docker` directory here?

> I'm not against using it in CI, as long as possibility to use local/external IMAP server is still there.

How would you currently make the CI to use an external IMAP server?

alecpl commented:
> How would you currently make the CI to use an external IMAP server?

Let's use docker in CI. I want possibility to run tests (manually) against an external IMAP server (and local greenmail).

As for docker. I don't care what the image we use or build it is coming from. I don't want docker related stuff outside of .ci directory.

pabzm committed:
Fix UI tests by waiting for element before using it

This only was a flaky problem only occurring sometimes.

pabzm committed:
Force a new IMAP connection in plugin tests, too

In other code the initial connection is forced. Doing this here, too,
fixes occasional problems with lost imap connections.

pabzm committed:
Make waiting for zipfile's content more robust

pabzm commented:
This setup works in the CI and in a container, and still allows to run the script locally as well as using an external IMAP-Server. Please review!

miaulalala reviewed (approved):
Nice!

alecpl commented on .github/workflows/tests.yml:
```diff
@@ -92,7 +77,7 @@ jobs:
 
       - name: Roundcube configuration
-        run: cp .github/config-test.inc.php config/config-test.inc.php
+        run: cp .ci/config-test.inc.php config/config-test.inc.php
```
You should call run_tests.sh on Windows too.

pabzm replied:
Good point, thank you!

But why can't? The script runs successfully on the Windows-runner, too.

alecpl commented on tests/Browser/TestCase.php:
```diff
@@ -38,7 +38,6 @@ protected function newBrowser($driver)
     public static function prepare(): void
     {
         static::startWebServer();
-        static::startChromeDriver();
```
Maybe `if (!getenv('WEBDRIVER_CONNECT_URL'))`...?

alecpl commented:
Looks good, almost there.

pabzm committed:
CI: Run tests from script on Windows, too

pabzm committed:
CI: Do start local chrome if no connect URL is given

alecpl commented:
Looks good, but can we move compose.yaml from root?

pabzm committed:
Move compose.yml to tests/

This way it's less easy mistaken as usable for running Roundcubemail in
production.

pabzm commented:
> can we move compose.yaml from root?

I moved it into `tests/`, is that ok?

alecpl commented:
`.ci` folder would be a better location.

alecpl commented:
Everything else is in .ci including docker images and run scripts.

pabzm committed:
Move compose.yml to .ci/

pabzm commented:
I don't really agree, but don't find it very important either and want this merged, so here you go. Can you merge it now?
